src/init/features/functions/index.ts

```diff
@@ -30,6 +22,104 @@ export async function doSetup(setup: any, config: any, options: Options) {
       ensure(projectId, "runtimeconfig.googleapis.com", "unused", true),
     ]);
   }
+  setup.functions = {};
+
+  if (!config.src.functions) {
+    // if functions have not been initialized yet
+    setup.config.functions = [];
+    return codebaseSetup(setup, config, false);
+  } else {
+    // if functions have already been initialized
+    // make sure config is validated and normalized
+    setup.config.functions = normalizeAndValidate(setup.config.functions);
+
+    logger.info();
+    logger.info("Detected " + clc.bold("existing codebase(s)."));
+    logger.info();
+
+    const choices = [
+      {
+        name: "Initialize",
+        value: "new",
+      },
+      {
+        name: "Re-initialize",
+        value: "reinit",
+      },
+    ];
+    const initOpt = await promptOnce({
+      type: "list",
+      message: "Would you like to initialize a new codebase, or reinitialize an existing one?",
+      default: "new",
+      choices,
+    });
+    return codebaseSetup(setup, config, initOpt === "reinit");
+  }
+}
+
+/**
+ *  User dialogue to set up configuration for functions codebase.
+ */
+async function codebaseSetup(setup: any, config: Config, reinit: boolean): Promise<any> {
+  if (reinit) {
+    const choices = setup.config.functions.map((cbconfig: any) => {
+      return {
+        name: cbconfig["codebase"],
+        value: cbconfig["codebase"],
+      };
+    });
+    const codebase = await promptOnce({
+      type: "list",
+      message: "Which codebase would you like to re-initialize?",
+      choices,
+    });
+    const cbconfig = configForCodebase(setup.config.functions, codebase);
+    _.set(setup, "functions.source", cbconfig.source);
```

Comment:
I don't have strong opinions, but several folks prefer we stop using lodash since it often doesn't add too much value.

This is equivalent to:

```
setup.functions.source = cbconfig.source;
```

right?

Also - I'm a bit confused since I thought `setup.config.functions` was an array of configs?

Reply:
I agree about the lodash thing... I would be perfectly happy to stop using it entirely.

So from my understanding of the previous developer's code, `setup.functions` keeps track of data needed for the rest of the functions initialization process. Things like `source`, `codebase`, `lint`, tracks configuration data in the `setup.functions` object that the next functions in the control flow can access to determine how to configure the new codebase.

 `setup.config.functions` refers to the new functions config object we are going to write to `firebase.json` as soon as we are done. So the two are distinct, but maybe `setup.functions` is not aptly named. Maybe something like `setup.options.functions` might be a little more explicit?

Reply:
ah `setup.config.functions` vs `setup.config`. My eyes failed me 😛